### PR TITLE
feat(proxy): add data proxy engine implementation & test

### PR DESCRIPTION
**Pull request by steebchen**

### CODE REVIEW

File: engine/proxy.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	if err != nil {
+		return fmt.Errorf("parse prisma string: %w", err)
+	}
+	e.apiKey = u.Query().Get("api_key")
+	if e.apiKey == "" {
+		return fmt.Errorf("could not parse api key from data proxy prisma connection string")
+	}
+
+	e.url = getCloudURI(u.Host, hash)
+	// TODO temp: remove this once prisma data proxy works with recent stable versions – right now only 3.0.1 works
```
Comment (br0kend) [inline engine/proxy.go:70]: Data Proxy backend supports anything above 2.29.0. I think a simple check of `binaries.PrismaVersion` would do here, before building a path based on it.

File: engine/proxy.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	// apiKey contains the parsed prisma data proxy api key from the connection string
+	apiKey string
+}
+
+func (e *DataProxyEngine) Connect() error {
+	// Example uri: https://aws-eu-west-1.prisma-data.com/2.26.0/412bf0a1742a576d699fbd5102a4f725557eff3992995f2e18febce128794961/
+	hash := hashSchema(e.Schema)
+	logger.Debug.Printf("local schema hash %s", hash)
+
+	logger.Debug.Printf("parsing connectino string from database url %s", e.schemaEnvVarName)
```
Comment (br0kend) [inline engine/proxy.go:54]: s/connectino/connection/
As much as I love `connectino`: it sounds like a small, Italian `connection`

File: engine/proxy.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	e.apiKey = u.Query().Get("api_key")
+	if e.apiKey == "" {
+		return fmt.Errorf("could not parse api key from data proxy prisma connection string")
+	}
+
+	e.url = getCloudURI(u.Host, hash)
+	// TODO temp: remove this once prisma data proxy works with recent stable versions – right now only 3.0.1 works
+	e.url = "https://" + path.Join(u.Host, "3.0.1", hash)
+	logger.Debug.Printf("using %s as remote URI", e.url)
+	if err := e.uploadSchema(context.Background()); err != nil {
```
Comment (br0kend) [inline engine/proxy.go:73]: From what I remember TS client uploads the schema when it needs to - aka, it tries to perform  the request first. 
I think this diff is correct for the use case of Go, where the server is likely to be longer lived, and unlikely to be a lambda.
Perhaps it'd be worth a comment somewhere noting the difference between the two clients.

File: engine/proxy.go (hunk 1 above)
Comment (steebchen) [reply engine/proxy.go:70]: It does not seem to work with 3.1.1, which is why I set it to 3.0.1. I get a 500 on the schema upload in 3.1.1

Comment (steebchen) [reply engine/proxy.go:70]: `http status code 404 with response {"EngineNotStarted":{"reason":"SchemaMissing"}}`

maybe something changed?

**Comment by matthewmueller** [comment]
Do we have any feature gating? 

In the JS client we require `previewFeatures = ["dataProxy"]`.

```prisma
generator db {
  provider        = "go run github.com/prisma/prisma-client-go"
  engineType      = "dataproxy"
  previewFeatures = ["dataProxy"]
}
```

**Comment by steebchen** [comment]
About the preview feature:

I don't really see the point of adding previewFeatures as the data proxy mode is already the only source of truth of whether to toggle or not, so it would not provide any additional precautions or isolation, it would just introduce more complexity. The only thing I can think of is that it would be more clear to the end-user that it's a preview feature.


About the env var:

That's not possible right now, I will add this toady,

**Comment by steebchen** [comment]
PRISMA_CLIENT_ENGINE_TYPE support is here – just couldn't add a test for it yet as unfortunately that requires some refactoring of the tests due to how the test clients are currently generated

File: engine/proxy.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+func (e *DataProxyEngine) Do(ctx context.Context, payload interface{}, into interface{}) error {
+	startReq := time.Now()
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("payload marshal: %w", err)
+	}
+
+	body, err := e.request(ctx, "POST", "/graphql", data)
+	if err != nil {
+		return fmt.Errorf("request failed: %w", err)
```
Comment (br0kend) [inline engine/proxy.go:112]: If this doesn't try to upload a schema on a SchemaMissing 404 error _somewhere_ when handling this error, then we'll have a pretty fragile client.

**Pull request merged**